go/multipooler/backup/backup.go

```diff
@@ -0,0 +1,165 @@
+}
+
+// Backup performs a backup on a specific shard
+func Backup(ctx context.Context, configPath, stanzaName string, opts BackupOptions) (*BackupResult, error) {
+	// Validation
+	if opts.Type == "" {
+		return nil, mterrors.New(mtrpcpb.Code_INVALID_ARGUMENT, "type is required")
+	}
+
+	// Type validation and mapping to pgbackrest types
+	// pgbackrest uses abbreviated types: full, diff, incr
+	pgBackRestType := ""
+	switch opts.Type {
+	case "full":
+		pgBackRestType = "full"
+	case "differential":
+		pgBackRestType = "diff"
+	case "incremental":
+		pgBackRestType = "incr"
+	default:
+		return nil, mterrors.New(mtrpcpb.Code_INVALID_ARGUMENT,
+			fmt.Sprintf("invalid backup type '%s': must be one of: full, differential, incremental", opts.Type))
+	}
+
+	// Validate required backup configuration
+	if configPath == "" {
+		return nil, mterrors.New(mtrpcpb.Code_INVALID_ARGUMENT, "config_path is required")
+	}
+	if stanzaName == "" {
+		return nil, mterrors.New(mtrpcpb.Code_INVALID_ARGUMENT, "stanza_name is required")
+	}
+
+	// Execute pgbackrest backup command
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Minute) // Backups can take a long time
+	defer cancel()
+
+	args := []string{
+		"--stanza=" + stanzaName,
+		"--config=" + configPath,
+		"--type=" + pgBackRestType,
+		"--log-level-console=info",
+	}
+
+	// Add annotations if table_group and shard are provided
+	if opts.TableGroup != "" {
+		args = append(args, "--annotation=table_group="+opts.TableGroup)
+	}
+	if opts.Shard != "" {
+		args = append(args, "--annotation=shard="+opts.Shard)
+	}
+
+	args = append(args, "backup")
+
+	cmd := exec.CommandContext(ctx, "pgbackrest", args...)
+
+	// Capture output for logging and to extract backup ID
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, mterrors.New(mtrpcpb.Code_INTERNAL,
+			fmt.Sprintf("pgbackrest backup failed: %v\nOutput: %s", err, string(output)))
+	}
+
+	// Parse the backup ID from the output
+	backupID, err := extractBackupID(string(output))
+	if err != nil {
+		return nil, mterrors.New(mtrpcpb.Code_INTERNAL,
+			fmt.Sprintf("failed to extract backup ID from output: %v\nOutput: %s", err, string(output)))
+	}
+
+	// Verify the backup to ensure it's valid
+	verifyCtx, verifyCancel := context.WithTimeout(context.Background(), 10*time.Minute)
+	defer verifyCancel()
+
+	verifyCmd := exec.CommandContext(verifyCtx, "pgbackrest",
+		"--stanza="+stanzaName,
+		"--config="+configPath,
+		"--set="+backupID,
+		"--log-level-console=info",
+		"verify")
+
+	verifyOutput, verifyErr := verifyCmd.CombinedOutput()
+	if verifyErr != nil {
+		return nil, mterrors.New(mtrpcpb.Code_INTERNAL,
+			fmt.Sprintf("pgbackrest verify failed for backup %s: %v\nOutput: %s", backupID, verifyErr, string(verifyOutput)))
+	}
+
+	return &BackupResult{
+		BackupID: backupID,
+	}, nil
+}
+
+// extractBackupID extracts the backup label from pgbackrest output
+//
+// TODO: find a way of of doing this that does that does not rely on text matching
```

Comment:
snag `pbackrest info --output=json` before and after? should be 'trivial' to parse out the new backup i think